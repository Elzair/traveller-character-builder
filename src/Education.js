import React from 'react';
import { checkMT2E, findTerm } from './utils';

export function Education({ 
    game, 
    career, 
    updateCareer, 
    upp, 
    updateUPP, 
    skills, 
    updateSkills,
    skillList,
    updateSkillList,
    age, 
    // education, 
    // updateEducation, 
    display, 
    onEducation, 
    updateLog 
}) {
    if (display && game === 'mt2e') {
        return (
            <EducationMT2E
                career={career}
                updateCareer={updateCareer}
                upp={upp}
                updateUPP={updateUPP}
                skills={skills}
                updateSkills={updateSkills}
                skillList={skillList}
                updateSkillList={updateSkillList}
                age={age}
                // education={education}
                // updateEducation={updateEducation}
                onEducation={onEducation}
                updateLog={updateLog}
            />
        );
    } else {
        return (<div></div>);
    }
}

// .##.....##.########..#######..########
// .###...###....##....##.....##.##......
// .####.####....##...........##.##......
// .##.###.##....##.....#######..######..
// .##.....##....##....##........##......
// .##.....##....##....##........##......
// .##.....##....##....#########.########

function EducationMT2E({ 
    career, 
    updateCareer, 
    upp, 
    updateUPP, 
    skills, 
    updateSkills, 
    skillList,
    updateSkillList,
    age, 
    // education, 
    // updateEducation, 
    onEducation, 
    updateLog 
}) {
    function educationDecision(ev) {
        ev.preventDefault();

        let decision = '';
        for (let t of ev.target) {
            if (t.checked) {
                decision = t.value;
            }
        }

        let roll;
        let newLog = [];
        let universityCareer = null;

        // Attempt entry into university
        if (decision === 'university') {
            // Traveller gets a -1 DM for each term after the first and a +1 if their Social is 9+.
            let dms = (findTerm(age)-1)*-1 + (upp.Social >= 9 ? 1 : 0);
            
            if (checkMT2E(7, upp, 'Education', null, null, dms).success) {
                newLog.push('Congratulations! You got into a university!');

                updateSkillList({
                    name: '',
                    list: [
                        'Admin',
                        'Advocate',
                        'Animals (Traning)',
                        'Animals (Veterinary)',
                        'Art (Performer)',
                        'Art (Holography)',
                        'Art (Instrument)',
                        'Art (Visual Media)',
                        'Art (Write)',
                        'Astrogation',
                        'Electronics (Comms)',
                        'Electronics (Computer',
                        'Electronics (Remote Ops)',
                        'Electronics (Sensors)',
                        'Engineer (M-drive)',
                        'Engineer (J-drive)',
                        'Engineer (Life Support)',
                        'Engineer (Power)',
                        'Language (Anglic)',
                        'Language (Vilani)',
                        'Language (Zdetl)',
                        'Language (Oynprith)',
                        'Medic',
                        'Navigation',
                        'Profession (Belter)',
                        'Profession (Biologicals)',
                        'Profession (Civil Engineering)',
                        'Profession (Construction)',
                        'Profession (Hydroponics)',
                        'Profession (Polymers)',
                        'Science (Archaeology)',
                        'Science (Astronomy)',
                        'Science (Biology)',
                        'Science (Chemistry)',
                        'Science (Cosmology)',
                        'Science (Cybernetics)',
                        'Science (Economics)',
                        'Science (Genetics)',
                        'Science (History)',
                        'Science (Linguistics)',
                        'Science (Philosophy)',
                        'Science (Physics)',
                        'Science (Planetology)',
                        'Science (Psionicology)',
                        'Science (Psychology)',
                        'Science (Robotics)',
                        'Science (Sophontology)',
                        'Science (Xenology)'
                    ],
                    values: [0, 1],
                    newSkills: {}
                });

                universityCareer = {
                    branch: 'university',
                    term: 0,
                    rank: null,
                    drafted: null,
                    rankPrev: null
                };

                // Increase Education by 1
                let newUPP = { ...upp };
                newUPP.Education += 1;
                updateUPP(newUPP);
            } else {
                newLog.push('Unfortunately, you did not qualify for university.');
                decision = 'career';
            }
        } else if (decision === 'army') {
            // Traveller gets a -2 DM for each term after the first one.
            let dms = (findTerm(age)-1)*-2;
            
            if (checkMT2E(8, upp, 'Endurance', null, null, dms).success) {
                universityCareer = {
                    branch: 'army-academy',
                    term: 0,
                    rank: null,
                    drafted: null,
                    rankPrev: null
                };

                newLog.push('Congratulations! You got into an army academy!');
            } else {
                newLog.push('Unfortunately, you did not qualify for an army academy.');
                decision = 'career';
            }
        } else if (decision === 'navy') {
            // Traveller gets a -2 DM for each term after the first one.
            let dms = (findTerm(age)-1)*-2;
            
            if (checkMT2E(9, upp, 'Intellect', null, null, dms).success) {
                universityCareer = {
                    branch: 'naval-academy',
                    term: 0,
                    rank: null,
                    drafted: null,
                    rankPrev: null
                };

                newLog.push('Congratulations! You got into a navy academy!');
            } else {
                newLog.push('Unfortunately, you did not qualify for a navy academy.');
                decision = 'career';
            }
        } else if (decision === 'marine') {
            // Traveller gets a -2 DM for each term after the first one.
            let dms = (findTerm(age)-1)*-2;
            
            if (checkMT2E(9, upp, 'Endurance', null, null, dms).success) {
                universityCareer = {
                    branch: 'marine-academy',
                    term: 0,
                    rank: null,
                    drafted: null,
                    rankPrev: null
                };

                newLog.push('Congratulations! You got into a marine academy!');
            } else {
                newLog.push('Unfortunately, you did not qualify for a marine academy.');
                decision = 'career';
            }
        }


        // Add university term to career list if traveller enrolled.
        if (universityCareer !== null) {
            let newCareer = [...career];
            newCareer.push(universityCareer);
            updateCareer(newCareer);
        }

        updateLog(newLog);
        onEducation(decision);
    }

    return (
        <form className="MT2EEducation" onSubmit={educationDecision}>
            <p>Do you want to further your education or start a career?</p>
            <label><input type="radio" id="education-university" name="education" value="university" />University</label>
            <label><input type="radio" id="education-army" name="education" value="army" />Army Academy</label>
            <label><input type="radio" id="education-navy" name="education" value="navy" />Navy Academy</label>
            <label><input type="radio" id="education-marine" name="education" value="marine" />Marine Academy</label>
            <label><input type="radio" id="education-career" name="education" value="career" />Start Career</label>
            <input className="Submit" type="submit" value="Submit" />
        </form>
    );
}