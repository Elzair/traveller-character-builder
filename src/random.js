var array = [];

export function r1d6() {
    return array.length > 0 ? array.shift() : (Math.floor(Math.random()*6)+1)
}

export function r2d6() {
    return r1d6() + r1d6();
}

export function randElt(arr) {
    return arr[array.length > 0 ? array.shift() : (Math.floor(Math.random()*arr.length))];
}

export function getValues(arr) {
    array = [...arr]
}