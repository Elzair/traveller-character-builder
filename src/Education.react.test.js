import React, { useEffect } from "react";
import ReactDOM from 'react-dom/client';
import * as reactTestRenderer from 'react-test-renderer';
// import { cleanup, fireEvent, render } from '@testing-library/react';
import { setValues } from './random';
import { Education } from "./Education";

//Education({ game, career, updateCareer, upp, updateUPP, skills, updateSkills, skillList, updateSkillList, age,  display, onEducation, updateLog })
test('Education renders nothing when `display` is false', () => {
    const game = 'mt2e';
    const career = [];
    const updateCareer = () => {};
    const upp = {Strength: 7, Dexterity: 7, Endurance: 7, Intellect: 7, Education: 7, Social: 7};
    const updateUPP = () => {};
    const skills = {};
    const updateSkills = () => {};
    const skillList = [];
    const updateSkillList = () => {};
    const age = 18;
    const cascadeSkill = null;
    const updateCascadeSkill = () => {};
    const display = false;
    const onEducation = () => {};
    const updateLog = () => {};

    const component = reactTestRenderer.create(
        <Education 
            game={game} 
            career={career} 
            updateCareer={updateCareer} 
            upp={upp} 
            updateUPP={updateUPP} 
            skills={skills} 
            updateSkills={updateSkills}
            skillList = {skillList}
            updateSkillList={updateSkillList}
            age={age}
            display={display}
            onEducation={onEducation}
            updateLog={updateLog} 
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('EducationMT2E accepts someone into university.', () => {
    setValues([5]);
    const game = 'mt2e';
    const career = [];
    const updateCareer = () => {};
    const upp = {Strength: 7, Dexterity: 7, Endurance: 7, Intellect: 7, Education: 12, Social: 7};
    const updateUPP = () => {};
    const skills = {};
    const updateSkills = () => {};
    const skillList = [];
    const updateSkillList = () => {};
    const age = 18;
    const cascadeSkill = null;
    const updateCascadeSkill = () => {};
    const display = true;
    let decision = null;
    const onEducation = dec => decision = dec;
    let log = [];
    const updateLog = updated => log = log.concat(updated);

    const {queryByLabelText, queryByText} = render(
        <Education 
            game={game} 
            career={career} 
            updateCareer={updateCareer} 
            upp={upp} 
            updateUPP={updateUPP} 
            skills={skills} 
            updateSkills={updateSkills}
            skillList = {skillList}
            updateSkillList={updateSkillList}
            age={age}
            display={display}
            onEducation={onEducation}
            updateLog={updateLog} 
        />
    );

    // const {queryByLabelText, queryByText} = render(
    //     <Education 
    //         game={game} 
    //         career={career} 
    //         updateCareer={updateCareer} 
    //         upp={upp} 
    //         updateUPP={updateUPP} 
    //         skills={skills} 
    //         updateSkills={updateSkills}
    //         skillList = {skillList}
    //         updateSkillList={updateSkillList}
    //         age={age}
    //         display={display}
    //         onEducation={onEducation}
    //         updateLog={updateLog} 
    //     />
    // );

    // // console.log(queryByText(/Submit/));
    
    // fireEvent.click(queryByLabelText(/University/));
    // fireEvent.click(queryByText(/Submit/));

    // expect(decision).toEqual('university');
    expect(career).toEqual(expect.arrayContaining([expect.objectContaining({ branch: 'university', term: 0, rank: null, drafted: null, rankPrev: null})]));
    expect(upp.Education).toEqual(13);
});