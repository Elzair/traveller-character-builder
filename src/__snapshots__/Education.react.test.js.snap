// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Education renders nothing when \`display\` is false 1`] = `<div />`;
